Add test for ordering when nils are present

jeff303 opened the pull request:
Adding test to ensure that we can handle a column containing nils in an ORDER BY clause

Adding `format-name` override for `bigquery-cloud-sdk` to convert dash to underscore in test table names

Adding comment showing IDs of each row in the `daily-bird-counts.edn` file

jeff303 committed:
Add test for ordering when nils are present

Adding test to ensure that we can handle a column containing nils in an ORDER BY clause

Adding `format-name` override for `bigquery-cloud-sdk` to convert dash to underscore in test table names

Adding comment showing IDs of each row in the daily-bird-counts.edn file

jeff303 committed:
Set Vertica sorts-nil-first? to true, since the test failure strongly implies it does

Format last column as int to make Oracle happy

Add impls for presto and presto-jdbc sorts-nil-first? as false

jeff303 committed:
More test fixes

Make the `sorts-nil-first?` multimethod accept a base type in order to return an answer

Vertica, at least, has a different answer for different types; see:
https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/AnalyzingData/Optimizations/NULLPlacementByAnalyticFunctions.htm#2

Updating all invocations of `sorts-nil-first?` in tests to pass the relevant base type

Updating Vertica implementation to consider the `base-type` as per its documentation

jeff303 commented:
To clarify, this PR is just adding a test (plus a bit of supporting code) to confirm that `ORDER BY` works properly even if the rows contain `NULL` values.  There isn't any type of bug fix or anything, just making sure something we expect to work, is working.  There are historically a few issues around this (see #6615 and linked items).  While those don't currently appear to be causing trouble, it's worth having a specific test for this to hopefully suss out any regressions going forward.

camsaul commented on modules/drivers/vertica/test/metabase/test/data/vertica.clj:
```diff
@@ -18,7 +18,10 @@
-(defmethod tx/sorts-nil-first? :vertica [_] false)
+;; In ORDER BY clause, nulls come last for FLOAT, STRING, and BOOLEAN columns, and first otherwise
+;; https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/AnalyzingData/Optimizations/NULLPlacementByAnalyticFunctions.htm#2
+(defmethod tx/sorts-nil-first? :vertica [_ base-type] (not (contains? #{:type/Text :type/Boolean :type/Float}
+                                                                      base-type)))
```
Probably worth having the body of this method on its own line

jeff303 replied:
Fixed

camsaul commented on test/metabase/test/util.clj:
```diff
@@ -997,3 +997,9 @@
       `(with-temp-file [])
       `(with-temp-file (+ 1 2)))))
+
+(defn dash->underscore
+  "Replaces dashes with underscores in `nm`. Intended for use in driver test data classes, to remap database, table,
+  etc. names in DBs that don't support having dashes in the identifier."
+  [nm]
+  (str/replace nm #"-" "_"))
```
We already have `metabase.util/snake-key` that does this. But we don't even really need to be implementing that ourselves since [`camel-snake-kebab`](https://github.com/clj-commons/camel-snake-kebab) is a thing

jeff303 replied:
As discussed on Slack, that library is also converting to lowercase in addition to snake casing.  So we will stick with our own impl for now.

jeff303 committed:
PR feedback

Add dependency on camel-snake-kebab so we can ditch our own impls

Fix Vertica method body

jeff303 committed:
PR feedback

Ditch camel-snake-kebab after all, since it also converts to lowercase
